src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs

```diff
@@ -414,16 +411,81 @@ public async Task RemoveAccountAsync(ApplicationConfiguration appConfig, IAccoun
             }
         }
 
-        public Task<IReadOnlyList<IAccount>> GetAccountsAsync(
+        public async Task<IReadOnlyList<IAccount>> GetAccountsAsync(
             string clientID,
             string redirectUri,
             AuthorityInfo authorityInfo,
             ICacheSessionManager cacheSessionManager,
             IInstanceDiscoveryManager instanceDiscoveryManager)
         {
-            // runtime does not yet support account discovery
+            if (!_wamOptions.ListWindowsWorkAndSchoolAccounts)
+            {
+                _logger.Info("[WamBroker] ListWindowsWorkAndSchoolAccounts option was not enabled.");
+                return Array.Empty<IAccount>();
+            }
+
+            Debug.Assert(s_lazyCore.Value != null, "Should not call this API if msal runtime init failed");
+
+            List<IAccount> msalAccounts = new List<IAccount>();
+            var requestContext = cacheSessionManager.RequestContext;
+
+            using (var discoverAccountsResult = await s_lazyCore.Value.DiscoverAccountsAsync(
+                clientID,
+                cacheSessionManager.RequestContext.CorrelationId.ToString("D"), 
+                cacheSessionManager.RequestContext.UserCancellationToken).ConfigureAwait(false))
+            {
+                if (discoverAccountsResult.IsSuccess)
+                {
+                    List<NativeInterop.Account> wamAccounts = discoverAccountsResult.Accounts;
+
+                    _logger.Verbose($"[WamBroker] Broker returned {wamAccounts.Count()} account(s).");
+
+                    if (wamAccounts.Count() == 0)
+                    {
+                        return Array.Empty<IAccount>();
+                    }
+
+                    //If "multi-cloud" is enabled, we do not have to do instanceMetadata matching
+                    if (!requestContext.ServiceBundle.Config.MultiCloudSupportEnabled)
+                    {
+                        var environmentList = discoverAccountsResult.Accounts.Select(acc => acc.Environment).Distinct().ToList();
+
+                        var instanceMetadata = await instanceDiscoveryManager.GetMetadataEntryTryAvoidNetworkAsync(
+                                authorityInfo,
+                                environmentList,
+                                requestContext).ConfigureAwait(false);
+
+                        _logger.Verbose($"[WamBroker] Filtering WAM accounts based on Environment.");
+
+                        wamAccounts.RemoveAll(acc => !instanceMetadata.Aliases.ContainsOrdinalIgnoreCase(acc.Environment));
+
+                        _logger.Verbose($"[WamBroker] {wamAccounts.Count} account(s) returned after filtering.");
+                    }
+                    
+                    foreach (var acc in wamAccounts)
+                    {
+                        msalAccounts.Add(WamAdapters.ConvertToMsalAccount(acc, clientID, _logger));
+                    }
+
+                    s_lazyCore.Value?.Dispose();
```

Comment:
Why do you dispose of the core object? I thought this remains alive for the duration of the process?